[DOCS 3242] serverless getting started (#13761)
DataDog/documentation: content/en/getting_started/serverless/_index.md

---
title: Getting Started with Serverless Monitoring
kind: documentation

further_reading:
    - link: '/agent/basic_agent_usage/'
      tag: 'Documentation'
      text: 'Basic Agent Usage'
---

## Overview

_Serverless_ is a model where developers build and run applications and services using a cloud provider, rather than managing infrastructure themselves. Datadog [Serverless Monitoring][10] collects metrics, logs, and traces from your serverless infrastructure, enabling you to monitor your application's health and performance.

> {{< img src="getting_started/serverless/serverless_view.png" alt="Serverless Monitoring: Serverless View, an explorer page" style="width:80%;">}}

Reviewer: Shall we use the sample app (instead of exposing org 2 data)?

Reviewer 2: +1

> This guide makes use of a serverless [sample app][1] that you can launch with one click. This app has Serverless Monitoring preconfigured. Follow this guide to see how you might troubleshoot a problem in your sample app, and what kinds of visibility Serverless Monitoring can provide.

Reviewer 2: ```suggestion
This guide uses a serverless [sample app][1] that you can launch with one click. This app has Serverless Monitoring preconfigured. Follow this guide to see how you might troubleshoot a problem in your sample app, and what kinds of visibility Serverless Monitoring can provide.
```

### Install the sample app

1. [Launch the CloudFormation Stack][9]. This link brings you to a **Create stack** page in CloudFormation.

> 2. Enter your Datadog API key and Datadog site ({{< region-param key="dd_site" code="true" >}}). 

Reviewer: I guess it doesn't hurt to link "Datadog API key" to the right page for convenience? Also for "Datadog site", lots of people don't change the site selector on the doc site when reading public docs, I think it would be nice (at least doesn't hurt) to also link it to the https://docs.datadoghq.com/getting_started/site.

  {{< img src="getting_started/serverless/aws_create_stack.png" alt="Close-up of two functions" style="width:80%;">}}

>   Then, acknowledge IAM capabilities and click **Create Stack**.

Reviewer: Looks like you need additional indentation here?

>   Then, acknowledge IAM capabilities and click **Create Stack**.

Reviewer 2: ```suggestion
    Then, acknowledge IAM capabilities and click **Create Stack**.
```

3. After the stack has been created, open the Outputs tab.

  {{< img src="getting_started/serverless/aws_outputs.png" alt="Close-up of two functions" style="width:80%;">}}

  Invoke your sample app a few times by visiting the `ApiGatewayInvokeURL`. This returns a "Sent message to SNS" success message.

> Each invocation executes the following:
> 
> ```python
> import boto3, os
> 
> def handler(event, context):
>     sns = boto3.client('sns')
> 
>     sns.publish(
>         TopicArn=os.environ.get("SNS_TOPIC_ARN"),
>         Message='Message sent to SNS'
>         )
> 
>     return {
>         "body": "Sent message to SNS",
>         "statusCode": 200
>     }
> ```
> 
> You can [see your sample app functions in Serverless View][11].
> 
> {{< img src="getting_started/serverless/dd_serverless_view.png" alt="Serverless Monitoring: Serverless View, an explorer page" style="width:80%;">}}

Reviewer 2: ```suggestion
    Each invocation executes the following:

    ```python
    import boto3, os

    def handler(event, context):
        sns = boto3.client('sns')

        sns.publish(
            TopicArn=os.environ.get("SNS_TOPIC_ARN"),
            Message='Message sent to SNS'
            )

        return {
            "body": "Sent message to SNS",
            "statusCode": 200
        }
    ```

    You can [see your sample app functions in Serverless View][11].

    {{< img src="getting_started/serverless/dd_serverless_view.png" alt="Serverless Monitoring: Serverless View, an explorer page" style="width:80%;">}}
```

make sure this suggestion works in the preview

## Serverless View

The Serverless View displays telemetry from all serverless resources in your AWS environment. You can use this page as a starting point for monitoring, debugging, and optimizing your applications.

If you have invoked your sample app at least once, you will see `datadog-sample-entry-function` and `datadog-sample-sqs-consumer-function`:

{{< img src="getting_started/serverless/functions_view.png" alt="Close-up of two functions" style="width:80%;">}}

> ### Serverless Insights

Reviewer: We can also highlight the ability to "group by" insights on the serverless view? Something like hey the default serverless view is grouped by service, but you can also change the "group by" and for instance group by "insights"? I think that looks very appealing for DevOps/SRE folks trying to identify problematic functions.

## Create an error to investigate

You can intentionally cause an error by editing the `datadog-sample-entry-function` in the sample app stack.

```python
  # Entry Lambda Function Code
  def handler(event, context):

    raise Exception('Throw an error.')
```

{{< img src="getting_started/serverless/aws_error.png" alt="Close-up of two functions" style="width:80%;">}}

Deploy this change and invoke your sample app again to see how you can investigate this error in Datadog.

{{< img src="getting_started/serverless/dd_serverless_view_error.png" alt="Close-up of two functions" style="width:80%;">}}

Notice that `datadog-sample-entry-function` has five errors.

## Function details
Click on your function to see more details regarding invocations and recent deployments.

{{< img src="getting_started/serverless/details_error.png" alt="Close-up of two functions" style="width:80%;">}}

The detailed view, as shown above, contains three graphs. You can set these to display any available metric; by default, they show three [enhanced Lambda metrics][7]: invocations, errors, and duration. 

> Datadog generates enhanced Lambda metrics out-of-the-box with low latency, several second granularity, and detailed metadata for cold starts and custom tags. You can also view the default [enhanced Lambda metrics dashboard][8].

Reviewer 2: ```suggestion
Datadog generates enhanced Lambda metrics out-of-the-box with low latency, several second granularities, and detailed metadata for cold starts and custom tags. You can also view the default [enhanced Lambda metrics dashboard][8].
```

### Invocations
The **Invocations** tab displays your function’s latest invocations. 

Each invocation is associated with a trace. Click on **Open Trace** to see the trace for each invocation:

{{< img src="getting_started/serverless/dd_flame_graph.png" alt="Close-up of two functions" style="width:80%;">}}

> 

Reviewer 2: Do you want a Flamegraph H3 here?

> The **Flame Graph** tab shows exactly what happened during the duration of this invocation, including which services had the highest percentage of the total execution time. The flame graph displays the request as it travels from APIGateway, through your `datadog-sample-entry-function`, through SNS, SQS, and finally your `datadog-sample-sqs-function`.

Reviewer 2: ```suggestion
The **Flame Graph** tab shows what happened during the duration of this invocation, including which services had the highest percentage of the total execution time. The flame graph displays the request as it travels from APIGateway, through your `datadog-sample-entry-function`, through SNS, SQS, and finally, your `datadog-sample-sqs-function`.
```

{{< img src="getting_started/serverless/trace_map.png" alt="Close-up of two functions" style="width:80%;">}}

> 

Reviewer: We should also highlight the request/response payload captured on the lambda span, which is a great help to troubleshooting errors.

The **Trace Map** tab visualizes the flow of your services and how they connect to each other.

The lower half of the detailed trace view displays a stack trace, which reveals the line of code responsible for throwing the error:

```
Traceback (most recent call last):
  File /opt/python/lib/python3.9/site-packages/datadog_lambda/wrapper.py, line 142, in __call__
    self.response = self.func(event, context, **kwargs)
File /var/task/index.py, line 17, in handler
    raise Exception('Throw an error.')
Exception: Throw an error.
```

> Underneath, you can also examine your Lambda request and response payloads. Datadog collects event payloads for every Lambda invocation.

Reviewer: I think it's worth including a screenshot of this most popular feature (request/response payload capturing).

### Logs

> The serverless sample app has logs enabled by default. You can see each function's logs under its **Logs** tab. 

Reviewer: I think it's more valuable to highlight (screenshot) the capability of seeing all the logs from one single invocation (clicking the invocation row)? We can mention the logs tab as well, but think that's very basic, just need a quick mentioning, but not worth being highlighted (compared to being able to see logs by invocations).

{{< img src="getting_started/serverless/dd_logs_view.png" alt="Close-up of two functions" style="width:80%;">}}

You can filter these logs to only see errors, or view them in the [Log Explorer][6].

> 

Reviewer: As we discussed previously, we should end this guide with showing prospects that they can easily set up a monitor and get a nice-looking alert in slack.
